### ISSUE TITLE
net/http/httputil: export function singleJoiningSlash

### ISSUE THREAD

**Issue by FabianKramm** [golang/go#44290]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.16beta1 darwin/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes, the function singleJoiningSlash is not exported.

https://github.com/golang/go/blob/1004a7cb31ae31d2ca0b54b507b996c12403d54c/src/net/http/httputil/reverseproxy.go#L102-L112

### What did you expect to see?

The function should be exported, since several open source projects have to copy it instead of using it directly, which is obviously bad:

Openshift:
https://github.com/openshift/console/blob/master/pkg/proxy/proxy.go#L69

AWS Proxy:
https://github.com/acquia/aws-proxy/blob/master/proxy/reverseproxy.go

Chronograf:
https://github.com/influxdata/influxdb/blob/master/chronograf/server/proxy.go

**Comment by neild** [comment]
> which is obviously bad

This is certainly not obvious to me. Copying a small, highly-specialized function seems much better than adding a dependency.

**Comment by FabianKramm** [comment]
@neild thanks for your response! In all mentioned repositories (which I only copied the first 3 I found on google and there seem to be a lot more that copy this function and import that package), this package is already a dependency, so you would not introduce a new one. 

Regardless of the question if code copying of small functions is bad, I don't really see a reason why making this function publicly accessible would cause any harm, when it has a clear benefit for people that do not want to copy code.

**Comment by neild** [comment]
Every piece of exported API surface comes with a cost, both for the package exporting it and for every package importing it. The costs are less obvious on the users' side, but are real: Increased API surface makes a package more difficult to use, any change to a dependency has the potential to break the code depending on it, the existence of an exported function to do something may incorrectly imply that this is a thing that should be done.

It is tempting to see an internal function in a package, see a use for it, and then say that the function should be exported because it is demonstrably useful. Following this approach causes package APIs to grow organically, not by design.

Package APIs should grow because the design suggests it, not just because the code is there.

**Comment by FabianKramm** [comment]
@neild I definitely agree that exporting functions always bears a risk and should correspond to a packages design, however I think in this case the design of the `net/http/httputil` package aligns very well with what the function is doing. According to the package itself the intend is to [provide HTTP utility functions, complementing the more common ones in the net/http package.](https://github.com/golang/go/blob/1004a7cb31ae31d2ca0b54b507b996c12403d54c/src/net/http/httputil/httputil.go#L5)

The function `singleJoiningSlash` is clearly a utility function for me that fits into this and is also clearly needed by the community to be consumed. It is also small, highly specialized and probably will never change, so I think the risk of exposing it is assessable. I don't agree that the best practice in this case should be to just copy internal package code if you already consume this package.

**Comment by toothrot** [comment]
I agree that the function has utility, but I disagree with "probably will never change". It's already been mentioned in several issues, and I can easily see how changing some corner cases (such as for empty strings, or similar) could be convenient for the package, but detrimental or impossible if it is exported.

See #25710 as an example of some hidden complexity.

**Comment by FabianKramm** [comment]
@toothrot I see thanks, I was not aware of the discussions around it. I still feel somewhat unsatisfied with this, because I see very big open source projects copying and relying on this code (including me) without any knowledge about the corner cases and hidden complexity around it and they all have to find out about this themselves. 

So while I agree that exporting the function as is might be suboptimal, is there not another way to provide a well documented publicly accessible function for that functionality in this package?

**Comment by networkimprov** [comment]
cc @ianlancetaylor as this seems like a Proposal.

**Comment by seankhliao** [comment]
I think this is covered by net/url.JoinPath
https://pkg.go.dev/net/url#JoinPath
